### PR TITLE
feat(sharing): add users in one request

### PR DESCRIPTION
**Pull request by audichuang**
## Summary

`SHARE_NOTEBOOK` already accepts many user entries in one call — the wire slot at index 1 of the
first param is a **list** of `[email, None, permission]` triples. `SharingAPI.add_user()` has always
put exactly one entry in it, so sharing a notebook with a team costs N RPC + status-refresh round
trips for what the backend will do in one. #2000 live-verified the plural shape on a scratch
notebook, including **mixed** permissions in a single call.

What batching saves is those round trips. It is **not** established to reduce how many notification
emails recipients receive: `notify` and `welcome_message` are per call, but nothing observed says
the backend sends fewer messages per grantee. An earlier revision of this description claimed "N
notification emails" — that claim was unfounded and has been removed.

This adds `SharingAPI.set_users()` for that case and re-expresses `add_user()` and `update_user()`
as 1-element calls of it, so there is a single validation path and a single wire-shape site rather
than several that can drift apart.

## Related Issue

Refs #2000 — **client half only**, so deliberately not a closing keyword. #2000 asks for plural
support in the MCP tool as well; `mcp/tools/sharing.py` is still singular and this PR does not touch
it (see Notes). Merging on `Closes` would auto-close the issue with that acceptance work untracked.

## Changes

- `src/notebooklm/_sharing.py`
  - New `set_users(notebook_id, grants, notify=True, welcome_message="")` where `grants` is a list of
    `(email, SharePermission)` pairs. It builds the entry list as
    `[[email, None, permission.value] for email, permission in grants]` — the shape #2000 recorded —
    sends one `SHARE_NOTEBOOK`, then refreshes `ShareStatus` once.
  - `add_user()` and `update_user()` now delegate to `set_users(notebook_id, [(email, permission)], ...)`. Their behaviour,
    signature, and its two `ValueError` messages are unchanged, and they are now raised from one
    place instead of two.
  - Validation runs before any RPC: empty `grants` raises `ValueError`, and `OWNER` / `_REMOVE`
    are rejected per grant with the pre-existing messages.
  - The DEBUG log keeps the grantee detail the single-user path used to carry. Delegating naively
    reduced `"Adding user %s … with permission %s"` to `"Adding 1 users to notebook nb_x"`, which drops
    *who* was granted *what* on the most common path — and that line's whole purpose is confirming a
    typoed address. It now logs the `(email, permission.name)` pairs for both the single and bulk
    callers.
- `docs/python-api.md` — `set_users` row in the `SharingAPI` table plus a worked mixed-permission
  example next to the existing `add_user` one.
- `docs/rpc-reference.md` — corrects the `SHARE_NOTEBOOK` section, which documented the entry slot as
  a single `[[email, None, permission]]`. It now shows a multi-entry `entries` list, states that
  `notify_flag` / `welcome_message` are **per call, not per user**, and records that `set_users()`
  issues one `SHARE_NOTEBOOK` + one `GET_SHARE_STATUS`. The web-control coverage table gains
  `set_users` alongside `add_user`.
- `tests/unit/test_sharing_types.py`
  - `TestSetUsers::test_set_users_sends_mixed_grants_in_one_request` decodes the outgoing `f.req`
    body and asserts the exact params, including both grants in one entry list, `notify_flag=0`, and
    `[message_flag=0, "Welcome, team!"]` — plus that exactly **two** requests are made
    (`SHARE_NOTEBOOK` then `GET_SHARE_STATUS`), which is the whole point of the method.
  - `test_set_users_rejects_invalid_grants` parametrises the `ValueError` cases (empty, OWNER,
    `_REMOVE`, duplicate) and asserts `httpx_mock.get_requests() == []`, so validation is pinned as
    pre-flight.
  - `test_set_users_upserts_an_existing_grantee`, `test_add_user_and_update_user_send_the_same_wire_shape`,
    `test_removal_keeps_its_own_message_block`, `test_case_variant_emails_are_not_treated_as_duplicates`,
    `test_set_users_notifies_by_default`, and `test_add_user_forwards_its_welcome_message`.
  - These are `httpx_mock` request-shape assertions, which `CONTRIBUTING.md` assigns to the **Unit**
    tier; they live there. (An earlier revision put them in `tests/integration/`; that was the
    subject of the now-resolved review thread.)
- `tests/integration/test_sharing_integration.py`
  - The existing `add_user` and `remove_user` tests gain a params assertion via the shared
    `rpc_request_params` fixture, so the single-entry wire shape is now pinned too — it previously
    was not, which is what let this slot's real arity go unnoticed. No new node IDs are added to the
    integration `allow_no_vcr` set.

## Test Plan

- [x] I tested these changes locally
- [x] Tests pass (`uv run pytest --cov=src/notebooklm --cov-report=term-missing --cov-fail-under=90`)
- [x] Linting and formatting pass (`uv run pre-commit run --all-files`)
- [x] Type checking passes (`uv run mypy src/notebooklm --ignore-missing-imports`)
- [x] If this PR changes architectural shape, an ADR has been added or updated — N/A, no shape change

Run against the canonical contributor install **plus the extras the CI test matrix adds**
(`--extra browser --extra dev --extra markdown --extra mcp --extra server --extra impersonate`).
The lean `browser+dev+markdown` install alone skips ~1,500 tests and lands at 83% coverage, i.e.
under the gate — so the suite below is the full-extras run that matches CI:

```console
$ uv run mypy src/notebooklm --ignore-missing-imports
Success: no issues found in 351 source files

$ uv run pre-commit run --all-files
ruff.....................................................................Passed
ruff-format..............................................................Passed

$ uv run pytest --cov=src/notebooklm --cov-report=term-missing --cov-fail-under=90 -q
TOTAL                                                    30937    826   8128    507    96%
Required test coverage of 90% reached. Total coverage: 96.41%
15384 passed, 64 skipped, 3 xfailed, 253 warnings in 565.52s (0:09:25)
```

I also ran every remaining gate from the `Code Quality` job of `.github/workflows/test.yml`, all green:
`check_workflow_permissions`, `check_workflow_secret_gates`, `check_action_pinning`,
`check_deprecation_targets`, `check_coverage_thresholds`, `check_ci_install_parity`,
`check_claude_md_freshness`, `check_docs_module_refs`,
`audit_public_api_compat.py --check-stale`, `tests/scripts/check_method_coverage.py`,
`pytest tests/e2e --collect-only`, and the lean-install core-import assertion
(`import notebooklm, notebooklm.notebooklm_cli, notebooklm.client, notebooklm.artifacts`).

## Notes

**Scope: this is the client half of #2000, not the MCP half.** #2000 also proposes a plural mode for
the MCP `share_set_user` tool (keeping its `confirm=true` gate, with the preview listing every
grantee). That is a separate surface with its own ADR-0025 schema-character budget and its own
confirmation-preview semantics, so it is left for a follow-up rather than bundled here. Nothing in
this PR blocks it — the MCP tool can adopt `set_users()` unchanged.

**No new idempotency-policy slot is needed.** #2000 mentions one, but `_idempotency_policy.py` is
keyed by `RPCMethod`, not by client method, and `RPCMethod.SHARE_NOTEBOOK` is already registered
(`NON_IDEMPOTENT_*` with the `GET_SHARE_STATUS` probe note). `set_users()` sends that same RPC, so it
inherits the existing classification rather than needing a new one.

**`notify` / `welcome_message` are per call.** The wire has one `notify_flag` and one
`[message_flag, message]` pair for the whole request, so a bulk call cannot vary them per grantee.
That is now stated in the docstring, in `docs/python-api.md`, and in `docs/rpc-reference.md` rather
than left for a caller to discover.

**Not verified live.** The multi-entry acceptance and the mixed-permission behaviour are from the
maintainer's live capture in #2000, not an independent run — I have no account positioned to
re-verify. My verification is offline: the params assertions above pin that we send exactly the shape
#2000 recorded. The Python 3.10–3.14 matrix is also unverified locally; the change uses no
version-sensitive syntax.

## Summary by CodeRabbit

- **New Features**
  - Added bulk notebook sharing for multiple recipients with mixed viewer and editor permissions in one operation.
  - Added notification and welcome-message options that apply to the whole call.
  - Single-recipient sharing now uses the same bulk-sharing capability.

- **Bug Fixes**
  - Added validation for empty recipient lists and unsupported permissions before requests are sent.

- **Documentation**
  - Documented the bulk-sharing API with usage examples and behavior details.

---

## Revision after the maintainer's live probe (2026-08-12)

@teng-lin probed this exact head against real accounts, and the results changed the public contract,
not just the wording. All six requested revisions are in.

**1. The operation is an upsert, so it is named one.** `add_users()` -> **`set_users()`**. The probe
showed a batch replaces an existing grantee's permission and adds an absent one, in both the plural
and singular forms. "Add" and "update" were never distinct operations, only distinct intents, so
`add_user()` and `update_user()` are now wrappers over `set_users()` differing only in their default
`notify`. No `update_users()` was added, per your note.

**2. Duplicate emails are rejected before any RPC.** A repeated grantee returns success while leaving
the permission unchanged, so there is no first-wins or last-wins rule to honour. `set_users()` raises
`ValueError` naming the offending address, and the test asserts `httpx_mock.get_requests() == []`.

The comparison is **exact**. I first made it case-insensitive and have reverted that: the probe
covered the same address twice, not case variants, and RFC 5321 keeps the mailbox local part
case-sensitive (only the domain is not), so folding case would have let the client hard-reject two
addresses the server may treat as distinct identities — a rule nobody has observed. `casefold()` also
collapses more than case (`straße@` and `strasse@` become one). Exact matching satisfies what was
actually probed; `test_case_variant_emails_are_not_treated_as_duplicates` pins the narrower behaviour,
and `docs/rpc-reference.md` flags the open question as worth a probe row if you run one again.

**3. No `remove_users()`.** Not added. `remove_user()`'s docstring now records *why* — the
all-or-nothing silent drop when any target is already absent — so the next reader does not take the
singular form for an oversight.

**4. Wire construction centralised, with one correction worth flagging.** A private `_share_params()`
now owns the `SHARE_NOTEBOOK` envelope for grants and removals alike; validation, logging, and the
status refresh live in `set_users()`; removal keeps its own explicit call site.

Doing this surfaced that the two paths **disagree on the message block**: the grant path sends
`[0, msg]` with a welcome message and `[1, ""]` without, while removal has always sent `[0, ""]`.
My first version derived that flag inside the encoder, which silently rewrote the removal payload
`test_remove_user` pins. `_share_params()` therefore takes `message_block` explicitly, and
`test_removal_keeps_its_own_message_block` pins **both** shapes so the discrepancy cannot be tidied
into a wire change later. It was previously invisible; it is now stated in the encoder's docstring.

**5. Contract and test language updated.**
- `docs/python-api.md`: upsert semantics, absent emails added, duplicate rejection, the per-call
  notify/welcome caveat, and why plural removal is absent.
- `docs/rpc-reference.md`: your eight probe rows recorded as an entry-list semantics table, plus the
  two consequences the client encodes.
- `CHANGELOG.md`: rewritten around upsert plus both preconditions.
- This description: the "N notification emails" claim is gone (see Summary).
- New tests: `test_set_users_upserts_an_existing_grantee` and
  `test_add_user_and_update_user_send_the_same_wire_shape`, which pins the two wrappers as one
  operation.

**6. Rebased** onto current `main`; the outdated test-tier thread is resolved.

### CODE REVIEW

**Comment by audichuang** [comment]
Submission was premature. The branch and worktree are retained for manual review by the user.

**Comment by teng-lin** [comment]
@codex review
@claude review

File: tests/integration/test_sharing_integration.py [hunk 1]
```diff
@@ -320,6 +344,100 @@ async def test_add_user_with_welcome_message(
         assert len(status.shared_users) == 2
 
 
+class TestAddUsers:
+    """Tests for SharingAPI.add_users()."""
```
[the remark this answers is not in the document]
Comment (audichuang) [reply tests/integration/test_sharing_integration.py]: Moved in 63f6bdc8 — the recommendation is right and I had already offered it in the PR description, so no argument. But the stated mechanism is not what fails, and that part matters for anyone reading this later.

**The suite was not failing.** `tests/_guardrails/test_integration_allow_no_vcr_allowlist.py` runs `scripts/test_taxonomy_inventory.py --check`, and `--check` / `--check-generated` are **mutually exclusive actions**:

```
--check            validate stable taxonomy policy
--check-generated  fail if generated allowlist baseline outputs differ
```

The baseline comparison is the second one, and it has **no call site** — not in `.github/workflows/`, not in `tests/`, not in tooling. The script's own generated header says so explicitly: *"`--check` validates the stable taxonomy policy only. This report and the exact node ID lists are audit aids; they are not a CI snapshot contract."* Consistent with CI on this PR being 19/19 green, and with `--check` passing locally (`OK: taxonomy policy is current`).

It also already reports stale on `origin/main`: `--check-generated` there lists both `integration_allow_no_vcr_files.txt` and `integration_allow_no_vcr_nodeids.txt`, so that staleness is pre-existing rather than introduced here.

**What I did anyway,** because the tier placement was genuinely wrong (CONTRIBUTING.md:141 puts httpx_mock request-shape assertions in Unit, and the grandfathered allowlist entry says "migrate per test-suite taxonomy cleanup"):

- `TestAddUsers` now lives in `tests/unit/test_sharing_types.py`, beside the existing `add_user` OWNER/`_REMOVE` validation coverage.
- The request-decode helper became an `rpc_request_params` fixture in `tests/conftest.py` instead of being duplicated — both tiers assert on wire shape and `test_no_cross_test_imports.py` forbids one test module importing another.
- The two assertions I added to the *existing* integration `add_user` / `remove_user` tests stay put (they add no node IDs) and use the shared fixture.

Net effect: this branch adds **zero** node IDs to the integration allow_no_vcr set. Full suite green on the rebase: 15223 passed, 96.57% coverage.

**Comment by teng-lin** [comment]
I ran a live probe against this exact head (`63f6bdc8`) on 2026-08-11, using two temporary scratch notebooks and the established test-account profiles from #2000. Both grantees were real test accounts, every mutation used `notify=False`, and both notebooks were deleted and verified absent afterward.

### Live results

| Operation | Observed behavior |
|---|---|
| Add two distinct users with mixed viewer/editor permissions | Succeeded; owner status showed both grants and both grantees could open the notebook directly |
| Call `add_users()` again for two existing users with flipped permissions | Succeeded; both permissions changed |
| Batch one existing-user update with one absent-user add | Succeeded |
| Call singular `update_user()` for an absent user | Added the user |
| Send duplicate/conflicting entries for the same email | RPC returned without error but the existing permission stayed unchanged — a silent no-op |
| Remove two users when both are present | Both were removed |
| Remove multiple users when any requested email is already absent | The entire request silently did nothing, even when the present user was first; unchanged after a follow-up status read |
| Remove one user and update a different user in the same request | Succeeded |

This establishes that the entry list is a general batch **set/upsert** operation for distinct emails, not specifically an add operation. It also exposes two backend preconditions that the current offline tests cannot show: duplicate grantees silently no-op, and multi-remove silently no-ops if any target is absent.

### Requested revision

1. **Settle the public naming around the real upsert semantics.**

   My preference is to name the new bulk primitive `set_users()`, `set_user_permissions()`, or `upsert_users()`. Existing `add_user()` and `update_user()` can remain compatibility/intent wrappers over one private encoder.

   If you retain `add_users()` for discoverability, document prominently that it also updates existing users. I would not add a semantically distinct `update_users()`: live behavior proves that “update” also adds an absent user, exactly like the existing singular wrapper.

2. **Reject duplicate emails before issuing any RPC.**

   Duplicate/conflicting entries for one email return success but leave the grant unchanged. Please validate uniqueness across `grants` and raise a clear `ValueError` before the request. Add a test asserting both the exception and `httpx_mock.get_requests() == []`. Rejecting every duplicate email is safer than attempting first/last-wins behavior the backend does not provide.

3. **Do not add a naive `remove_users()` in this PR.**

   Multiple permission-4 entries work only when every target currently exists. If any email is already absent, present users are not removed and the RPC provides no failure signal. A reliable plural removal needs a separate design:

   - preflight `GET_SHARE_STATUS`;
   - intersect requested emails with currently shared users;
   - return immediately if none remain;
   - batch-remove the present set;
   - refresh and verify;
   - retry/reconcile or fall back to singular removals if a race leaves requested users behind.

   That keeps the singular operation’s practical “remove if present” semantics instead of exposing a silent all-or-nothing trap.

4. **Centralize wire construction, but keep removal reconciliation separate.**

   A private `_set_users(...)`/entry-encoding helper can own validation, logging, the `SHARE_NOTEBOOK` envelope, and the final status refresh for add/update wrappers. A future `remove_users()` may reuse the encoder, but its preflight/postcondition logic should remain explicit.

5. **Update the contract and test language.**

   - Document that grants are upserts and that absent emails are added.
   - Document duplicate rejection.
   - Keep the per-call notification/welcome-message caveat.
   - Correct the PR description’s “N notification emails” claim: batching removes N RPC/status-refresh cycles, but does not establish that fewer recipient emails are sent.
   - Record the live duplicate and removal preconditions in the RPC reference or a linked follow-up so they are not rediscovered.

6. **Rebase onto current `main`.**

   After #2178 merged, this branch now has a `CHANGELOG.md` conflict. Please rebase after the API revision, rerun the full checks, and resolve the already-addressed outdated test-tier thread.

The bulk grant implementation itself is live-confirmed. The blockers are the public upsert contract and duplicate-grantee silent no-op; `remove_users()` should be a separately designed follow-up rather than a mechanical plural wrapper.

**Comment by audichuang** [comment]
Rebased onto current `main` (66385599) — this was `DIRTY`, now mergeable.

The only conflict was `CHANGELOG.md`: both this branch and the merged #2141 prepend to the same `### Added` slot. Both entries kept, new one on top, matching how every prior PR has landed there. `docs/python-api.md` and `docs/rpc-reference.md` auto-merged.

Suite green after the rebase: 1915 passed, 3 xfailed across `tests/unit/test_sharing_types.py`, `tests/integration/test_sharing_integration.py`, and `tests/_guardrails`.

@claude's two non-blocking observations are noted and deliberately not acted on here: duplicate emails within one `grants` list are not deduped (the backend's tie-break is unverified, and guessing at it client-side would bake in an assumption this PR can't test), and the DEBUG log's list comprehension is eager (grants lists are team-sized; making it lazy costs more than it saves). Happy to take either as a follow-up if you'd rather have them.

**Comment by audichuang** [comment]
Pushed `d2613743`. All six of your revisions are in, and an adversarial review pass on top of them found four things I had got wrong — those are fixed too, and one of them is a reversal of something I told you I was doing, so it leads.

### Reversed: duplicate detection is exact, not case-insensitive

My first pass folded case, and I flagged it here as a judgement call. It was the wrong call and I have reverted it.

Your probe repeated the **same** address; it never covered case variants, so I was encoding an unobserved rule. And the direction is not the safe one: [RFC 5321 §2.3.11](https://www.rfc-editor.org/rfc/rfc5321.html#section-2.3.11) keeps the mailbox local part case-sensitive — only the domain is not — so folding case lets the *client* hard-reject two addresses the server may treat as distinct identities. `casefold()` also collapses more than case: `straße@example.com` and `strasse@example.com` become one key.

Exact matching covers exactly what you probed. `test_case_variant_emails_are_not_treated_as_duplicates` pins the narrower behaviour so a future "improvement" back to `casefold()` has to bring evidence, and `docs/rpc-reference.md` records the open question as worth a probe row if you run one again.

### Two missing guards, both caught by mutation

Neither of these was visible from reading the diff — I found them by breaking the source and watching the suite stay green:

- **`add_user()` dropping `welcome_message` in its delegation** left all 72 tests passing. It now has `test_add_user_forwards_its_welcome_message`.
- **Flipping `set_users()`'s `notify` default from `True` to `False`** also left all 72 passing — every other test passes `notify` explicitly, and the stable-API baseline does not cover a brand-new method's defaults. Now `test_set_users_notifies_by_default`.

Both fail against the broken version; I re-ran the mutations after adding them to confirm.

### `Closes #2000` → `Refs #2000`

#2000 also asks for a plural mode on the MCP `share_set_user` tool. `mcp/tools/sharing.py` is still singular and this PR does not touch it, so merging on `Closes` would auto-close the issue with that acceptance work untracked. Same correction you asked for on #2179, applied here for the same reason.

### The six revisions

1. **`add_users()` → `set_users()`**, with `add_user()` / `update_user()` as intent wrappers differing only in their default `notify`. No `update_users()`. `test_add_user_and_update_user_send_the_same_wire_shape` pins the two wrappers as one operation so they cannot re-split.
2. **Duplicates rejected pre-flight** with a `ValueError` naming the address; the test asserts `httpx_mock.get_requests() == []`.
3. **No `remove_users()`.** `remove_user()`'s docstring now records *why* — the all-or-nothing silent drop — so the singular form does not read as an oversight.
4. **`_share_params()` owns the envelope** for grants and removals; validation, logging and the status refresh stay in `set_users()`; removal keeps its explicit call site.

   One thing this surfaced, worth your eye: **the two paths disagree on the message block.** Grants send `[0, msg]` with a welcome message and `[1, ""]` without; removal has always sent `[0, ""]`. My first version derived that flag inside the encoder, which silently rewrote the removal payload — `test_remove_user` caught it. `_share_params()` therefore takes `message_block` explicitly, and `test_removal_keeps_its_own_message_block` pins **both** shapes. Whether removal's `[0, ""]` is meaningful or just historical, I do not know; it is at least no longer invisible.
5. **Contract and test language** updated across `docs/python-api.md`, `docs/rpc-reference.md` (your eight probe rows recorded verbatim), `CHANGELOG.md`, and this PR description — including dropping the unfounded "N notification emails" claim. Batching saves N RPC + status-refresh round trips; nothing observed says recipients get fewer emails.
6. **Rebased** onto current `main`; the outdated test-tier thread is resolved.

### Verification

Full suite green: **15384 passed, 64 skipped, 3 xfailed** (565s). `mypy` clean, `ruff` lint + format clean. The review pass independently re-derived the payloads for `add_user` (with and without a message), `update_user`, `remove_user`, `set_public` and `set_view_level` against the pre-change code and confirmed all six are byte-identical — `set_public` / `set_view_level` were untouched and stay that way. `grep add_users` returns nothing; the public-surface audit collects `set_users` dynamically, so no baseline needed editing.

**Comment by teng-lin** [comment]
@claude review
@codex review
